### PR TITLE
feat: Payments based dunning

### LINKED ISSUE #14842: Finance/Service Charge added automatically
In our industry and many others, a service charge is added to a customer's account automatically upon no payment after a certain time period. Typically, it's net 30 or net 15 so if their invoice isn't paid in those terms, a % charge is added to the invoice. We have always issued this as a new invoice of sorts - we named it SC#####. This invoice was auto-created in our old system based on the terms. We would run a report of all service charges monthly and occasionally, we might go in and cancel or remove one if there was an error or problem with shipment, pricing dispute, etc..  but the SC invoice was created automatically for us. This would be a nice feature to have. 

Almost all of our customers incur service charges but we do have a few select ones that do not so having a check box on the customer doctype for "Service Charges" is necessary. For us, all of our service charges are the same 1.5% of the invoice past due balance.

I am willing to pay for this feature.

Thanks!
Heather

### PR DESCRIPTION
**Pull request by barredterra**
Related to #26969

Test this PR at https://gh-imp-2021-01268.castlecraft.co.in (User: mossdiana@hotmail.com, Password: 7b97830487)

Create dunning for all overdue payments of a customer. We use a table similar to the payment schedule in sales invoice to pull all overdue scheduled payments into the Dunning.

- Rate of Interest is the same for all overdue payments.
- Interest amount is calculated for each overdue payment.
- Dunning Fee is applied only once per dunning.

### Changes in accounting

- Dunning amount will no longer be booked to receivables. The reason is that the dunning amount is not paid in many cases / is only intended as a warning. If paid, the dunning amount will be recorded as additional income in Payment Entry. See [here](https://www.rechnungswesenforum.de/forum/buchfuehrung-buchhaltung/allgemein/forderung-mahngebuehren.423727/) and [here](https://www.haufe.de/finance/haufe-finance-office-premium/mahnung-und-mahnverfahren-73-buchung-mahngebuehren-und-verzugszinsen_idesk_PI20354_HI2015327.html) for further reading (in german).

Preview:

new_dunning

Docs: https://docs.erpnext.com/docs/v13/user/manual/en/accounts/dunning/edit?wiki_page_patch=894eeb58e3
Discuss: https://discuss.erpnext.com/t/dunning-mahnung-warning-notice/49499/30?u=rmeyer

To do:

- [x] Refactor `resolve_dunning`
- [x] Add patch
- [x] Adjust docs
- [x] Update test cases to use new mandatory **Company** field in **Dunning Type**.

### CODE REVIEW

**Comment by barredterra** [comment]
@KanchanChauhan I'd appreciate your feedback on this.

**Comment by KanchanChauhan** [comment]
@barredterra Looks good to me. But I was wondering if instead of pulling all the outstanding Sales Invoices, we have a pop-over window like shown below where the user selects which invoices he/she wants to select. Some invoices might be for the first Reminder and some for second, in that case, the Rate of Interests will be different and will require separate Dunning.

Screen Shot 2021-09-16 at 10 14 32 AM

**Comment by barredterra** [comment]
@KanchanChauhan thanks for the hint, now I'm using this feature to let the user fetch overdue payments from specific invoices.

I think the interest rate always stays the same, but the first reminder might not have any interest charged.

**Comment by barredterra** [comment]
@mergifyio backport version-13-hotfix

**Comment by barredterra** [comment]
Failing patch seems to be unrelated

**Comment by barredterra** [comment]
@KanchanChauhan please check if this PR is okay for your use-case and let me know if any changes are necessary.

Test site: https://discuss.erpnext.com/t/bot-discussion-pr-27528-feat-payments-based-dunning/81598

**Comment by HarryPaulo** [comment]
> Resolved ✅ 

image

target.income_account = dunning_type.income_account
target.cost_center = dunning_type.cost_center

On create a dunning from sales invoice, need to load the icome accout and cost center from dunning type

**Comment by HarryPaulo** [comment]
> Resolved ✅ 

The button to create the dunning does not appear in the following situation:

Payment schedule:
Due date         |      rate
2021-11-15     |     100.00 <<<<<----- due payment
2021-12-15     |     100.00

Today, November 17, the first registration must allow you to create a dunning. 

Improvement suggestion:
1 - remove the if "if (doc.due date < frappe.datetime.get_today()) {"
image

2 - Validate as expired when generating the dunning 

image

"condition": lambda doc: doc.outstanding > 0 and getdate(doc.due_date) < getdate(nowdate())


**With my last comment on this PR, this fix probably will be different **

**Comment by HarryPaulo** [comment]
> Resolved ✅ 

image

precision for field interest is not found.

right code:

precision("interest", row)

**Comment by HarryPaulo** [comment]
If i have two records on payments templates with due date, i generate a dunning

Example:

image

PS: i got customizations on this, it's not equals this PR, but this situation is equals for the PR and for me.

when i create a payment for this dunning, i get a message notify me about duplicated records for the same sales invoice: 

image

what i think:

Payment amout must be a sum of the dunning, on my situation is a sum of the allocated amout. on ur case is a sum of the outstanding of the dunning, and insert only **ONE** record for the sales invoice.

Remember: a dunning could containing many invoice from the same client, the group by must be per invoice.

**Comment by HarryPaulo** [comment]
With a **Sales Invoice** with many records on the payment schedule, when you create a payment to pay the first record from the payment schedule, you will pay partially, but the field `paid_amout` only update if both fields are set.

image

In some cases this can cause a problem, because will load again the payment schedule to the dunning, but it was paid and the values from payment schedule was not updated.

Idk if this is a problem, but  on **Payment Terms Template** `allocate_payment_based_on_payment_terms` must be enabled, and on table **payment term** field `payment_term` must be set. Then it will update the payment schedule when creating a payment.

**Alternative 1:**
When generating a **Dunning**, consider the parameter `allocate_payment_based_on_payment_terms`. If it is disabled, consider the amount paid on the **Sales Invoice**, allocate the paid amount on each payment schedule, sorting by due date.

**Alternative 2** (i think this is not the best alternative):
Payment Entry, update the payment schedule values.

**Alternative 3**: what do you think about it??

File: erpnext/accounts/doctype/dunning_type/dunning_type.py [hunk 1]
```diff
@@ -2,9 +2,11 @@
...
-# import frappe
+import frappe
 from frappe.model.document import Document
 
 
 class DunningType(Document):
-	pass
+	def autoname(self):
+		company_abbr = frappe.get_value("Company", self.company, "abbr")
+		self.name = self.dunning_type + " - " + company_abbr
```
Comment (hrwX) [inline erpnext/accounts/doctype/dunning_type/dunning_type.py]:
```suggestion
		self.name = f"{self.dunning_type} - {company_abbr}"
```

File: erpnext/accounts/doctype/dunning/dunning.js [hunk 2]
```diff
@@ -40,42 +51,112 @@ frappe.ui.form.on("Dunning", {
...
+					if (r.message) {
+						frm.set_value("company_address", r.message);
+					} else {
+						frm.set_value("company_address", "");
+					}
+				}
+			});
+
+			if (frm.fields_dict.currency) {
+				var company_currency = erpnext.get_currency(frm.doc.company);
```
Comment (hrwX) [inline erpnext/accounts/doctype/dunning/dunning.js]:
```suggestion
				const company_currency = erpnext.get_currency(frm.doc.company);
```

File: erpnext/accounts/doctype/dunning/dunning.js [hunk 3]
```diff
@@ -40,42 +51,112 @@ frappe.ui.form.on("Dunning", {
...
+					frm.set_value("currency", company_currency);
+				}
+
+				if (frm.doc.currency == company_currency) {
+					frm.set_value("conversion_rate", 1.0);
 				}
 			}
-		);
+
+			var company_doc = frappe.get_doc(":Company", frm.doc.company);
```
Comment (hrwX) [inline erpnext/accounts/doctype/dunning/dunning.js]:
```suggestion
			const company_doc = frappe.get_doc(":Company", frm.doc.company);
```

File: erpnext/accounts/doctype/dunning/dunning.js [hunk 4]
```diff
@@ -40,42 +51,112 @@ frappe.ui.form.on("Dunning", {
...
+			if (company_doc.default_letter_head) {
+				if (frm.fields_dict.letter_head) {
+					frm.set_value("letter_head", company_doc.default_letter_head);
+				}
+			}
+		}
+	},
+	currency: function (frm) {
+		// this.set_dynamic_labels();
+		var company_currency = erpnext.get_currency(frm.doc.company);
```
Comment (hrwX) [inline erpnext/accounts/doctype/dunning/dunning.js]:
```suggestion
		const company_currency = erpnext.get_currency(frm.doc.company);
```

File: erpnext/accounts/doctype/dunning/dunning.js [hunk 5]
```diff
@@ -40,42 +51,112 @@ frappe.ui.form.on("Dunning", {
...
+			frappe.call({
+				method: "erpnext.setup.doctype.company.company.get_default_company_address",
+				args: { name: frm.doc.company, existing_address: frm.doc.company_address || "" },
+				debounce: 2000,
+				callback: function (r) {
+					if (r.message) {
+						frm.set_value("company_address", r.message);
+					} else {
+						frm.set_value("company_address", "");
+					}
```
Comment (hrwX) [inline erpnext/accounts/doctype/dunning/dunning.js]:
```suggestion
					frm.set_value("company_address", r && r.message || "");
```

**Comment by deepeshgarg007** [comment]
@barredterra Closing this since its in an unmergeable state right now, please resolved conflicts and then reopen

**Comment by daanlenaerts** [comment]
@barredterra I landed here through the [Dunning docs](https://docs.erpnext.com/docs/v13/user/manual/en/accounts/dunning), which don't seem to match how dunnings are implemented in ERPNext.
It is not clear to me what is wrong exactly, but is there any prospect of this feature being merged? It looks extremely useful and much better than the currently available dunning implementation. Can I be of help?

**Comment by barredterra** [comment]
@daanlenaerts currently we don't have a sponsor for this feature, that's why work on it was suspended. Feel free to take over development yourself, or let me know if you'd like to sponsor.

As for the docs, it looks like the update meant for this PR got published prematurely.

**Comment by daanlenaerts** [comment]
@barredterra Thank you very much for your swift reply. I am considering sponsoring or continuing development of this feature, as I'd otherwise have to build something like this from the ground up, which would be rather unfortunate. Do you have an idea what would be required to finish this feature or how sponsorship would work?

**Comment by michalls01** [comment]
@barredterra Hello, Could you please let me know aprox. how much work is left on this PR? This is a very helpful extension! We might try to collaborate or sponsor on this one if possible. Thanks!

**Comment by barredterra** [comment]
@michalls01 As far as I remember, the following still needs to be done for this PR:

- [ ] Resolve conflicts
- [ ] Manual tests of functionality
- [ ] Test of migration patch,
- [ ] Possibly add some unit tests
- [ ] Apply changes suggested by reviewers / community
- [ ] Check and update the docs

Feel free to reach out to hallo@alyf.de for sponsoring.

**Comment by Libermentix** [comment]
@barredterra We are happy to sponsor this feature for the most recent version of ERPNext v14!

**Pull request closed without merging**